### PR TITLE
Find MPS ground state with pytorch autograd

### PR DESCRIPTION
**Pull request by benschiff**
Hi you can have a look in the bin folder :)

### CODE REVIEW

**Comment by GiggleLiu** [comment]
Nice, this looks very nice, you are really quick in learning a new tool and I am very sure you will become a very good Julia programmer! I will take a look.
👍🏾

**Comment by GiggleLiu** [comment]
I pushed a commit to your branch, normally we do not do this (since it will cause conflicts).
But I want to merge the main branch first and resolve conflicts with `git merge main` and remove `pyc` files so that I can review easier.

File: bin/pytorchMPSevolve.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
+import numpy as np
```
Comment (GiggleLiu) [inline bin/pytorchMPSevolve.py:1]: I would suggest removing numpy dependency because we do not really need it.

File: bin/pytorchMPSevolve.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+import torch
+from torch.autograd import Variable
+
+def init_spinup_MPS(N, D=1):
+    """MPS: all spins up
+    Args
+    ----
+    D : physical dimension, product state for D=1
+    """
+    Ms = np.zeros([N, D, 2, D], dtype=np.csingle)
```
Comment (GiggleLiu) [inline bin/pytorchMPSevolve.py:11]:
```suggestion
    Ms = torch.zeros([N, D, 2, D], dtype=torch.float32)
```

Then `Ms.requires_grad = True` to make it differentiable.

File: bin/pytorchMPSevolve.py [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+class MPS:
+    """Class for handling matrix product states assuming a Vidal form"""
+    def __init__(self, Ms):
+        self.Ms = Ms
+        self.N = len(Ms)
+
+    def get_theta1(self, i):
+        """Effective single-site wave function on site i"""
+        return torch.tensordot(torch.diag(self.Ss[i]), self.Ms[i], [[1], [0]])
```
Comment (GiggleLiu) [inline bin/pytorchMPSevolve.py:33]: Is `tensordot` faster than `einsum`? If not (very likely), I would suggest using einsum everywhere. Because it is much more readable.

File: bin/pytorchMPSevolve.py [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+if __name__ == "__main__":
+    N = 5
+    D = 1
+    print(f"+++ Ising chain with N={N}, MPS bond dim D={D} +++")
+    print(f"Finding the ground state with PyTorch autograd...\n")
+    psi = init_spinup_MPS(N, D)
+    model = IsingModel(N)
+    loss = model.energy(psi)
+    loss.backward()
+    optim = torch.optim.SGD([psi.Ms], lr=1e-2, momentum=0.9)
```
Comment (GiggleLiu) [inline bin/pytorchMPSevolve.py:135]: Try better optimizers like LBFGS, SGD is a very bad optimizer.

File: bin/pytorchMPSevolve.py (hunk 3 above)
Comment (benschiff) [reply bin/pytorchMPSevolve.py:33]: Aren't `einsum` and `tensordot` calling the same dot products under the hood? But I agree that `einsum` is nice.

**Comment by GiggleLiu** [comment]
@maddiecain You should review this PR, it is really a good first example for learning tensor networks!

File: bin/pytorchMPSevolve.py (hunk 3 above)
Comment (GiggleLiu) [reply bin/pytorchMPSevolve.py:33]: I think so, I am thinking about readability. Here, you can not see input tensor dimension for example.

Comment (GiggleLiu) [reply bin/pytorchMPSevolve.py:33]: FYI, here you should write `einsum("i,ij->ij", self.Ss[i], self.Ms[i])`

File: bin/pytorchMPSevolve.py (hunk 4 above)
Comment (benschiff) [reply bin/pytorchMPSevolve.py:142]: I can't get LBFGS to work at the moment, a first implementation is included in the .py script.

File: bin/PEPS.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+    def conj(self):
+        for i in range(self.N):
+            self.vertex_tensors[i] = self.vertex_tensors[i].conj()
+        return self
+
+    def output_variables(self):
+        return torch.cat([t.flatten() for t in self.vertex_tensors])
+
+    def load_variables(self, variables):
```
Comment (GiggleLiu) [inline bin/PEPS.py:52]: pytorch has its own parameter management system, you'd better use its own. It is available in `Module` system, for an example,
https://github.com/TensorBFS/dTRG/blob/312ff05cd40b3a3f7b0237b878d6f3275dd054d1/dTRG.py#L65

File: bin/PEPS.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    def get_einsum_order(self):
+        """ There is a problem oe.contract_expression, need to check. Workaround with on the fly contraction
+        order for now"""
+        # Statetensor
+        sublist = []
+        for i,t in enumerate(self.vertex_tensors):
+            sublist.append(self.vertex_labels[i])
+            sublist.append(t.shape)
+        print(sublist)
+        self.code_statetensor = oe.contract_expression(*sublist)
```
Comment (GiggleLiu) [inline bin/PEPS.py:25]: Didn't know that, learnt from you! 👍🏾

File: bin/PEPS.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+            sublist.append(peps2.vertex_tensors[i].conj())
+            t_c_labels = self.vertex_labels[i].copy() + largest_label
+            t_c_labels[-1] = self.vertex_labels[i][-1]
+            sublist.append(t_c_labels)
+        return oe.contract(*sublist)
+
+    def apply_Hamiltonian(self, H):
+        for h in H:
+            t, i = h
+            self.vertex_tensors[i] = oe.contract('...i,ij->...j', self.vertex_tensors[i], t)
```
Comment (GiggleLiu) [inline bin/PEPS.py:92]: Physical label is always the last dimension right? For the design perspective, I would suggest using a more explicit convention. Imagine your student wants to read your code.

File: bin/timeevolve.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+    pl.load_variables(variables)
+    pl.normalize()
+    pr = peps
+    pr.load_variables(variables)
+    pr.normalize()
+    return torch.abs(pr.overlap_with(pl))
+
+def smatrix(peps):
+    variables = peps.output_variables()
+    return hessian(lambda x: loss1(peps, x), variables)
```
Comment (GiggleLiu) [inline bin/timeevolve.py:26]: We do not need hessian, let's discuss!

**Review by GiggleLiu — COMMENTED** [review]
Good job! As you mentioned you don't know if your code is correct, I think a easy way to check is writing unit tests, it is suuuuuuper important. See e.g. this blog https://realpython.com/pytest-python-testing/

File: bin/timeevolve.py (hunk 8 above)
Comment (GiggleLiu) [reply bin/timeevolve.py:26]: https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.gmres.html

Comment (GiggleLiu) [reply bin/timeevolve.py:26]: https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html
